### PR TITLE
Revamped docker compose

### LINKED ISSUE #3306: Make the docker compose file self documenting
### It should be made clear in the docker compose file:
- The prerequisites required to run the file
- How to actually run the file
- The purpose for each of the environment variables used by different containers
- Their recommended/suggested/default values

### Goals:
- The need of referring to our docs for this installation method by the user should be eliminated upon resolving this issue.   
- The need of user to add variables in the `.env` file should be removed and instead these variables should be customizable in the compose file itself.

### Nice to haves:
- Make possible mounting folders as volumes to increase the visibility of the data being stored on user device. This is also useful if the data is being used by some other service on the host device. 
- Consider removing watchtower and push user to update manually instead.

### Additional context

Related to #3166

### PR DESCRIPTION
**Pull request by Anish9901**
Fixes #3306

**For the reviewer:** 
- Not adding any technical details here as the docker compose file should be self explanatory.
- Please go through the `docker-compose.yml` file, read through the comments and try setting up the prod version of Mathesar.
- Please report any feedback and/or difficulty you faced during the installation. 

## Checklist
 
- [x] My pull request has a descriptive title (not a vague title like `Update
  index.md`).
- [x] My pull request targets the `develop` branch of the repository
- [x] My commit messages follow [best practices][best_practices].
- [x] My code follows the established code style of the repository.
- [ ] I added tests for the changes I made (if applicable).
- [x] I added or updated documentation (if applicable).
- [x] I tried running the project locally and verified that there are no
  visible errors.

[best_practices]:https://gist.github.com/robertpainsi/b632364184e70900af4ab688decf6f53

## Developer Certificate of Origin

<details>
<summary>Developer Certificate of Origin</summary>

```
Developer Certificate of Origin
Version 1.1

Copyright (C) 2004, 2006 The Linux Foundation and its contributors.
1 Letterman Drive
Suite D4700
San Francisco, CA, 94129

Everyone is permitted to copy and distribute verbatim copies of this
license document, but changing it is not allowed.


Developer's Certificate of Origin 1.1

By making a contribution to this project, I certify that:

(a) The contribution was created in whole or in part by me and I
    have the right to submit it under the open source license
    indicated in the file; or

(b) The contribution is based upon previous work that, to the best
    of my knowledge, is covered under an appropriate open source
    license and I have the right under that license to submit that
    work with modifications, whether created in whole or in part
    by me, under the same open source license (unless I am
    permitted to submit under a different license), as indicated
    in the file; or

(c) The contribution was provided directly to me by some other
    person who certified (a), (b) or (c) and I have not modified
    it.

(d) I understand and agree that this project and the contribution
    are public and that a record of the contribution (including all
    personal information I submit with it, including my sign-off) is
    maintained indefinitely and may be redistributed consistent with
    this project or the open source license(s) involved.
```

</details>

### CODE REVIEW

File: docker-compose.yml [hunk 1]
```diff
@@ -1,100 +1,258 @@
...
+#-------------------------------------------------------
+
+
+#--------------------------------------------------
+# CONFIG:
+#--------------------------------------------------
+x-config: &config
+  # (Required) Replace '?' with '-' followed by your 50 character long Django
+  # SECRET_KEY, you can generate one here: https://djecrety.ir/
+  SECRET_KEY: ${SECRET_KEY:?}
```
Comment (mathemancer) [inline docker-compose.yml:22]: I'm debating whether we should include a default key that's obviously unfit for production. Something like:


```suggestion
  SECRET_KEY: ${SECRET_KEY:-PLEASEDONOTUSETHISSECRETKEYITISSUPERDUPERINSECURE}
```

**Review by mathemancer — COMMENTED** [review]
Overall, this is really great. Nice work, @Anish9901 !

I made some changes to get the unified docker image to work. I'd like for you to look at those before we have others work through this.

File: docker-compose.yml (hunk 1 above)
Comment (mathemancer) [reply docker-compose.yml:22]: After some discussion in various channels, I think the current failure state is better in this use-case.

**Comment by mathemancer** [comment]
@kgodey For testing, you need to run 

```
docker build -t mathesar/mathesar-prod:latest --build-arg PYTHON_REQUIREMENTS=requirements-prod.txt .
docker build -t mathesar/mathesar-caddy:latest -f Dockerfile.caddy .
```
from the root of the repo. This is because we haven't yet published the new images for 0.1.4. Actual users won't need to do that.

Then, just copy the docker compose file to any directory you want to use for the test installation, and follow the instructions in that file. I recommend using some directory other than the repo to avoid contamination from any dev files (e.g., `.env`) in the repo, and since this setup adds subdirectories under wherever it's run (as documented in the file).

File: run.sh [hunk 2]
```diff
@@ -2,10 +2,18 @@
...
+then
   echo "Starting inbuilt database"
   ./db-run.sh
-  export DJANGO_DATABASE_URL='postgres://postgres:mathesar@localhost:5432/mathesar_django'
+  export POSTGRES_USER=postgres
+  export POSTGRES_PASSWORD=mathesar
+  export POSTGRES_HOST=localhost
+  export POSTGRES_PORT=5432
+  export POSTGRES_DB=mathesar_django
   if [[ -z "${MATHESAR_DATABASES}" ]]; then
```
Comment (Anish9901) [inline run.sh:17]: @mathemancer this script fails when all the `POSTGRES_*` variables are explicitly provided but `MATHESAR_DATABASES` isn't.

Here is the command I ran:
```bash
docker run \
-e POSTGRES_DB=mathesar_django \
-e POSTGRES_HOST=localhost \
-e POSTGRES_PORT=5432 \
-e POSTGRES_USER=mathesar \
-e POSTGRES_PASSWORD='mathesar' \
-v static:/code/static \
-v media:/code/media \
-v postgresql_config:/etc/postgresql/ \
-v postgresql_data:/var/lib/postgresql/ \
--name mathesar_service  \
-p 8000:8000 \
--restart unless-stopped \
mathesar/mathesar-prod:latest
```

**Comment by seancolsen** [comment]
I've unassigned @mathemancer and assigned @kgodey per this [comment](https://matrix.to/#/!kpLRqiyQwsndKhBIvd:matrix.mathesar.org/$l_Ic-jPX0pIFV6DzbBBj8NH01AdSEqQHvm1ewynuBeI?via=matrix.mathesar.org&via=matrix.org) from @mathemancer

File: run.sh (hunk 2 above)
Comment (mathemancer) [reply run.sh:17]: Good catch; I've pushed a commit to fix it.

Comment (Anish9901) [reply run.sh:17]: Works perfectly after the fix, Brent!

**Comment by mathemancer** [comment]
After chatting in other channels, I'm going to merge this, and we'll handle issues in future PRs.

**Pull request merged**